core/src/ancestor_hashes_service.rs

```diff
@@ -290,86 +354,791 @@ impl AncestorHashesService {
+            request_throttle.len(),
+            ancestor_hashes_request_statuses.len()
+        );
+
+        // 5) If we've reached the throttle limit, no requests should be made,
+        // but should still read off the channel for replay updates
+        request_throttle.clear();
+        request_throttle.resize(MAX_ANCESTOR_HASHES_SLOT_REQUESTS_PER_SECOND, std::u64::MAX);
+        let dead_duplicate_confirmed_slot_2 = 15;
+        ancestor_hashes_replay_update_sender
+            .send(AncestorHashesReplayUpdate::DeadDuplicateConfirmed(
+                dead_duplicate_confirmed_slot_2,
+            ))
+            .unwrap();
+        AncestorHashesService::manage_ancestor_requests(
+            &ancestor_hashes_request_statuses,
+            &ancestor_hashes_request_socket,
+            &repair_info,
+            &outstanding_requests,
+            &ancestor_hashes_replay_update_receiver,
+            &serve_repair,
+            &mut AncestorRepairRequestsStats::default(),
+            &mut dead_slot_pool,
+            &mut repairable_dead_slot_pool,
+            &mut request_throttle,
+        );
+
+        assert_eq!(dead_slot_pool.len(), 1);
+        assert!(dead_slot_pool.contains(&dead_slot));
+        assert_eq!(repairable_dead_slot_pool.len(), 1);
+        assert!(repairable_dead_slot_pool.contains(&dead_duplicate_confirmed_slot_2));
+        assert_eq!(ancestor_hashes_request_statuses.len(), 1);
+        assert!(ancestor_hashes_request_statuses.contains_key(&dead_duplicate_confirmed_slot));
+
+        // 6) If root moves past slot, should remove it from all state
+        let bank_forks = &repair_info.bank_forks;
+        let root_bank = bank_forks.read().unwrap().root_bank();
+        let new_root_slot = dead_duplicate_confirmed_slot_2 + 1;
+        let new_root_bank = Bank::new_from_parent(&root_bank, &Pubkey::default(), new_root_slot);
+        new_root_bank.freeze();
+        {
+            let mut w_bank_forks = bank_forks.write().unwrap();
+            w_bank_forks.insert(new_root_bank);
+            w_bank_forks.set_root(new_root_slot, &AbsRequestSender::default(), None);
+        }
+        assert!(!dead_slot_pool.is_empty());
+        assert!(!repairable_dead_slot_pool.is_empty());
+        assert!(!ancestor_hashes_request_statuses.is_empty());
+        request_throttle.clear();
+        AncestorHashesService::manage_ancestor_requests(
+            &ancestor_hashes_request_statuses,
+            &ancestor_hashes_request_socket,
+            &repair_info,
+            &outstanding_requests,
+            &ancestor_hashes_replay_update_receiver,
+            &serve_repair,
+            &mut AncestorRepairRequestsStats::default(),
+            &mut dead_slot_pool,
+            &mut repairable_dead_slot_pool,
+            &mut request_throttle,
+        );
+        assert!(dead_slot_pool.is_empty());
+        assert!(repairable_dead_slot_pool.is_empty());
+        assert!(ancestor_hashes_request_statuses.is_empty());
+    }
```

Comment:
Can we get a case where we get a dead duplicate confirmed so we make a request. In the fake blockstore we dump this slot (because repair service will try to dump & repair at the same time), and verify that this causes no issues when processing the ancestor responses we get?

Reply:
Added test `test_ancestor_hashes_service_manage_ancestor_hashes_after_replay_dump`: https://github.com/solana-labs/solana/pull/18912/files#diff-c4189090057f79f53d5760d5a5214e68a8c73f22bb431a7c17829ab802903eeeR1312-R1313